gerritlab/global_vars.py

```diff
@@ -27,28 +31,143 @@ def load_config(remote, repo):
+    `load_config` function.
+    """
+    if "GITLAB_PRIVATE_TOKEN" in os.environ:
+        return os.environ["GITLAB_PRIVATE_TOKEN"]
+
+    # Try to get the private token from git config.
+    try:
+        return git_config.get_value("gerritlab", "private-token")
+    except (configparser.NoSectionError, configparser.NoOptionError):
+        pass
+
+    # Try to get the private token from git credentials.
+    private_token = GitCredentialStore(host).get_token()
+    if private_token is not None:
+        return private_token
+
+    # DEPRECATED: Try to get the private token from .gitreview file.
+    if "private_token" in gitreview_config:
+        raise RuntimeError("Using private_token from .gitreview file is deprecated.")
+
+    raise SystemExit(f"Unable to find private token for {host}")
+
+
+class GitCredentialStore:
+    """
+    A class to interact with git credential store.
+    """
+    def __init__(self, host: str) -> None:
+        # TODO: use host + repo path to support per-repo tokens
+        self._host = host
+        self._token: Optional[str] = None
+        self._git_credentials: Dict[str, str] = {}
+        self._git_credentials_fetched: bool = False
+
+    def get_token(self) -> Union[str, None]:
+        """
+        Convenience method to get the token from the credential store.
+        """
+        return self.get("password")
+
+    def get(self, key: str, default=None) -> Union[str, None]:
+        """
+        Return key from internal git credentials cache, or default if not found.
+        """
+        if not self._git_credentials_fetched:
+            self._populate_git_credentials(self._call_git_credential_fill())
+            self._git_credentials_fetched = True
+
+        return self._git_credentials.get(key, default)
+
+    def _populate_git_credentials(self, git_credentials_output: str) -> None:
+        """
+        Populate the internal git credentials cache
+        """
+        for line in git_credentials_output.splitlines():
+            if "=" in line:
+                key, value = line.split("=", 1)
+                self._git_credentials[key] = value
+
+    def _call_git_credential_fill(self) -> str:
+        """
+        Look up credentials for the host using git credential fill.
+
+        Prepend 'gerritlab@' to ensure gitlab tokens can be stored separately
+        from other user tokens if desired.
+        """
+        try:
+            git_credentials = subprocess.run(
+                ["git", "credential", "fill"],
+                input=f"protocol=gerritlab\nhost=gerritlab@{self._host}\n\n",
+                text=True,
+                capture_output=True,
+            )
+        except KeyboardInterrupt:
+            return ""
+
+        if git_credentials.returncode != 0:
+            return ""
+        git_credential_output = git_credentials.stdout
+
+        # Store the credentials for later use
+        self._call_git_credential_approve(git_credential_output)
```

Comment:
We shouldn't approve until we know that the credentials really worked.

Reply:
Tried an implementation of that in the latest revision, thoughts?